File: backend/app/schemas/audit_legacy.py
"""Public contract for checking uploaded workbooks, never importing their rows."""

from datetime import datetime, timezone
from typing import Annotated, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator


AuditLegacyKind = Literal["cases", "atoms", "assignments", "events", "daily_totals"]
AuditLegacyField = Literal[
    "case_key", "digital_product", "contract_date", "atom_key", "title",
    "source_clause", "work_type", "object_type", "state", "alpha_result",
    "alpha_date", "commission_result", "commission_date", "system_url",
    "assignee_email", "actor_name", "assigned_at", "is_current", "occurred_at",
    "event_key", "event_type", "metric_date", "metric_type", "value",
    "workflow_stage", "contract_reference", "source_evidence_text", "notes",
    "alpha_comment", "previous_state", "ended_at", "assignment_key",
]
ColumnLetter = Annotated[str, Field(pattern=r"^[A-Z]{1,3}$")]


class AuditLegacyColumn(BaseModel):
    column: ColumnLetter
    label: str


class AuditLegacySuggestedMapping(BaseModel):
    kind: AuditLegacyKind
    fields: dict[AuditLegacyField, ColumnLetter]


class AuditLegacySheet(BaseModel):
    id: str
    name: str
    row_count: int = Field(ge=0)
    column_count: int = Field(ge=0)
    formula_count: int = Field(ge=0)
    header_row: int = Field(ge=1)
    columns: list[AuditLegacyColumn]
    suggested_mapping: AuditLegacySuggestedMapping


class AuditLegacyInspection(BaseModel):
    parser_version: str
    sheets: list[AuditLegacySheet]


class AuditLegacyMapping(BaseModel):
    model_config = ConfigDict(extra="forbid")

    sheet_id: str = Field(min_length=1, max_length=512)
    header_row: int = Field(ge=1, le=1048576, strict=True)
    kind: AuditLegacyKind
    fields: dict[AuditLegacyField, ColumnLetter] = Field(max_length=35)

    @field_validator("fields")
    @classmethod
    def valid_columns(cls, fields: dict) -> dict:
        for column in fields.values():
            number = 0
            for letter in column:
                number = number * 26 + ord(letter) - ord("A") + 1
            if number > 16384:
                raise ValueError("Column is outside the XLSX range")
        return fields


class AuditLegacyMappingUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    revision: int = Field(ge=1, strict=True)
    mapping: AuditLegacyMapping


class AuditLegacyIssue(BaseModel):
    row: int | None
    column: str | None
    code: str
    severity: Literal["error", "warning"]
    message: str


class AuditLegacyPreviewRow(BaseModel):
    row: int
    values: dict[str, str | None]


class AuditLegacyReport(BaseModel):
    sheet_id: str
    header_row: int
    kind: AuditLegacyKind
    total_rows: int = Field(ge=0)
    valid_rows: int = Field(ge=0)
    error_rows: int = Field(ge=0)
    warning_rows: int = Field(ge=0)
    duplicate_rows: int = Field(ge=0)
    issues: list[AuditLegacyIssue]
    issue_count: int = Field(ge=0)
    preview_rows: list[AuditLegacyPreviewRow]
    columns: list[AuditLegacyColumn]
    unmapped_columns: list[str]
    ready_for_import: Literal[False]


class AuditLegacyBatchRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    sha256: str
    size_bytes: int
    status: Literal["uploaded", "checked"]
    created_at: datetime
    updated_at: datetime
    revision: int
    inspection: AuditLegacyInspection
    mapping: AuditLegacyMapping | None
    report: AuditLegacyReport | None

    @field_validator("created_at", "updated_at")
    @classmethod
    def utc_timestamp(cls, value: datetime) -> datetime:
        if value.tzinfo is None:
            value = value.replace(tzinfo=timezone.utc)
        return value.astimezone(timezone.utc)

File: frontend/src/api/auditLegacy.ts
import { api } from './client'

export type LegacyKind = 'cases' | 'atoms' | 'assignments' | 'events' | 'daily_totals'
export interface LegacyColumn { column: string; label: string }
export interface LegacyMapping {
  sheet_id: string
  header_row: number
  kind: LegacyKind
  fields: Record<string, string>
}
export interface LegacySheet {
  id: string
  name: string
  row_count: number
  column_count: number
  formula_count: number
  header_row: number
  columns: LegacyColumn[]
  suggested_mapping: { kind: LegacyKind; fields: Record<string, string> }
}
export interface LegacyReport {
  sheet_id: string
  header_row: number
  kind: LegacyKind
  total_rows: number
  valid_rows: number
  error_rows: number
  warning_rows: number
  duplicate_rows: number
  issues: Array<{ row: number | null; column: string | null; code: string; severity: 'error' | 'warning'; message: string }>
  issue_count: number
  preview_rows: Array<{ row: number; values: Record<string, string | null> }>
  columns: LegacyColumn[]
  unmapped_columns: string[]
  ready_for_import: false
}
export interface LegacyBatch {
  id: string
  sha256: string
  size_bytes: number
  status: 'uploaded' | 'checked'
  created_at: string
  updated_at: string
  revision: number
  inspection: { parser_version: string; sheets: LegacySheet[] }
  mapping: LegacyMapping | null
  report: LegacyReport | null
}

export const LEGACY_MAX_FILE_BYTES = 10 * 1024 * 1024
export const LEGACY_KINDS: Record<LegacyKind, string> = {
  cases: 'Договоры', atoms: 'Атомы', assignments: 'Назначения', events: 'События', daily_totals: 'Дневные итоги',
}
export const LEGACY_FIELDS = {
  case_key: 'Код договора', digital_product: 'Цифровой продукт', contract_date: 'Дата договора',
  atom_key: 'Код атома', title: 'Название атома', source_clause: 'Пункт источника', work_type: 'Вид работы',
  object_type: 'Тип объекта', state: 'Статус', alpha_result: 'Результат альфа-проверки', alpha_date: 'Дата альфа-проверки',
  commission_result: 'Результат комиссии', commission_date: 'Дата комиссии', system_url: 'Ссылка на систему',
  assignee_email: 'Email ответственного', actor_name: 'Имя участника', assigned_at: 'Дата назначения',
  is_current: 'Текущее назначение', occurred_at: 'Дата события', event_key: 'Код события', event_type: 'Тип события',
  metric_date: 'Дата показателя', metric_type: 'Тип показателя', value: 'Значение',
} as const
export type LegacyField = keyof typeof LEGACY_FIELDS
export const LEGACY_REQUIRED: Record<LegacyKind, LegacyField[]> = {
  cases: ['case_key', 'digital_product'],
  atoms: ['case_key', 'atom_key', 'title'],
  assignments: ['case_key', 'assignee_email', 'assigned_at'],
  events: ['case_key', 'event_key', 'event_type', 'occurred_at'],
  daily_totals: ['case_key', 'metric_date', 'metric_type', 'value'],
}

const root = '/api/audit/legacy-imports'
const batchPath = (id: string) => `${root}/${encodeURIComponent(id)}`
export const auditLegacy = {
  list: (signal?: AbortSignal) => api.get<LegacyBatch[]>(root, undefined, { signal }),
  get: (id: string, signal?: AbortSignal) => api.get<LegacyBatch>(batchPath(id), undefined, { signal }),
  upload: (file: File) => {
    const data = new FormData()
    data.append('file', file)
    return api.upload<LegacyBatch>(root, data)
  },
  check: (id: string, revision: number, mapping: LegacyMapping) =>
    api.put<LegacyBatch>(`${batchPath(id)}/mapping`, { revision, mapping }),
  source: (id: string) => api.blob(`${batchPath(id)}/source`),
  remove: (id: string, revision: number) => api.delete<void>(`${batchPath(id)}?revision=${revision}`),
}
